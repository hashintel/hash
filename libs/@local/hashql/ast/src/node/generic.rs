use hashql_core::{span::SpanId, symbol::Ident};

use super::{id::NodeId, r#type::Type};
use crate::heap;

/// A concrete type argument for a generic type.
///
/// Represents a specific type used to instantiate a generic type parameter.
/// In a generic type instantiation like `Vec<String>`, `String` would be
/// represented as a [`GenericArgument`].
///
/// # Examples
///
/// In a type reference like `HashMap<K, V>`, `K` and `V` are generic arguments.
/// In an instantiated type like `List<Int>`, `Int` is a generic argument.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct GenericArgument {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
}

/// A generic type parameter declaration.
///
/// Represents a type parameter in a generic type or function declaration.
/// Each parameter has a name and an optional type bound that constrains
/// what types can be used as arguments for this parameter.
///
/// # Examples
///
/// In a declaration like `fn<T>(x: T): T`, `T` is a generic parameter.
/// In a declaration like `fn<T: Number>(x: T): T`, `T` is a generic parameter
/// with a bound of `Number`.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct GenericParam<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
    pub bound: Option<heap::Box<'heap, Type<'heap>>>,
}

/// A collection of generic parameters for a type or function.
///
/// Represents the complete set of type parameters in a generic declaration,
/// such as the `<T, U>` in `fn<T, U>(x: T, y: U): U`.
///
/// # Examples
///
/// For a function like `fn<T, U: Comparable>(x: T, y: U): Bool`:
/// - The `Generics` would contain two `GenericParam` entries
/// - One for `T` with no bound
/// - One for `U` with a bound of `Comparable`
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Generics<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub params: heap::Box<'heap, [GenericParam<'heap>]>,
}
