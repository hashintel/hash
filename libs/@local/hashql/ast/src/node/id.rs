use hashql_core::id;

id::newtype!(
    /// A unique identifier for nodes in the HashQL Abstract Syntax Tree.
    ///
    /// Each node in the AST has a unique `NodeId` that can be used to track the node
    /// through various processing stages. These IDs are particularly important for
    /// error reporting, cross-referencing between nodes, and maintaining node identity
    /// during transformations.
    ///
    /// The value space is restricted to 0..=0xFFFF_FF00, reserving the last 256 for niches.
    /// As real pattern types are an experimental feature in Rust, these can currently only be
    /// used by directly modifying and accessing the `NodeId`'s internal value.
    ///
    /// # Implementation Note
    ///
    /// We're currently liberal in the use of `NodeId`s. In the future, we might want to
    /// revisit this to see if we can use fewer of them or optimize their allocation.
    pub struct NodeId(u32 is 0..=0xFFFF_FF00)
);

impl NodeId {
    /// A placeholder ID used during initial parsing.
    ///
    /// When parsing, we initially give all AST nodes this placeholder ID. After resolving special
    /// forms, the nodes are renumbered to have small indices.
    pub const PLACEHOLDER: Self = <Self as id::Id>::MAX;
}
