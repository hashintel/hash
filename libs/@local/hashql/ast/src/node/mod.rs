//! Core Abstract Syntax Tree node types for HashQL.
//!
//! This module defines the fundamental building blocks of the HashQL Abstract
//! Syntax Tree (AST). The AST represents the structure of a HashQL program after
//! parsing, capturing all the language constructs in a hierarchical tree structure.
//!
//! ## Module Organization
//!
//! The node module is organized into several submodules:
//!
//! - [`expr`]: Expression nodes representing computations that produce values
//! - [`generic`]: Generic type parameter and argument definitions
//! - [`id`]: Node identifier types and utilities
//! - [`path`]: Path expressions for variable references and qualified names
//! - [`type`]: Type system representations including struct, tuple, and other types
//!
//! ## Memory Management
//!
//! All AST nodes use arena allocation through the `'heap` lifetime parameter,
//! which references the memory region where nodes are allocated. This approach
//! improves performance by minimizing allocations and simplifies memory management
//! during parsing and analysis.
//!
//! ## Node Identification
//!
//! Each node in the AST has a unique [`id::NodeId`] that can be used to track
//! the node through various processing stages. Nodes also include a span identifier that points to
//! the source location of the node.
//!
//! ## AST Structure
//!
//! The AST is structured as a tree with expressions ([`expr::Expr`]) at the core.
//! Other node types like types ([`type::Type`]) and paths ([`path::Path`]) are
//! used as components within expressions or on their own for declarations.
//!
//! The entire tree represents a complete HashQL program or query, ready for
//! analysis, transformation, and evaluation.
pub mod expr;
pub mod generic;
pub mod id;
pub mod path;
pub mod r#type;
