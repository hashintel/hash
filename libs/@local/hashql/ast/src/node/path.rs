use hashql_core::{span::SpanId, symbol::Ident};

use super::{generic::GenericArgument, id::NodeId};
use crate::heap;

/// A segment in a path expression.
///
/// Represents a single component of a qualified path, consisting of an identifier
/// and optional generic type arguments. Path segments are connected to form
/// complete paths like `module::submodule::item<T>`.
///
/// # Examples
///
/// In a path like `std::collections::HashMap<K, V>`:
/// - `std` is a segment with no generic arguments
/// - `collections` is a segment with no generic arguments
/// - `HashMap` is a segment with generic arguments `K` and `V`
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct PathSegment<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
    /// Type parameters attached to this path segment
    pub arguments: heap::Vec<'heap, GenericArgument<'heap>>,
}

/// A path expression in the HashQL Abstract Syntax Tree.
///
/// Represents a qualified reference to a variable, function, type, or module.
/// Paths can be simple identifiers (like `x`) or qualified names with multiple
/// segments separated by double colons (like `std::collections::HashMap`).
///
/// Path expressions are used for variable references, module access, and type names.
/// They can include generic type arguments applied to specific segments.
///
/// # Examples
///
/// Simple paths:
/// ```text
/// x                     // A local variable
/// length                // A function or variable
/// ```
///
/// Qualified paths:
/// ```text
/// std::collections::HashMap  // An item in a nested module
/// math::sin                  // A function in a module
/// ```
///
/// Paths with generic arguments:
/// ```text
/// Vec<Int>                       // A type with one type argument
/// HashMap<String, User>          // A type with two type arguments
/// graph::query<User, Connection> // A generic function
/// ```
///
/// J-Expr explicit:
/// ```json
/// {"#path", ["std", "collections", "HashMap"]}
/// {"#path", [{"name": "HashMap", "arguments": ["K, V"]}]}
/// ```
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Path<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    /// Whether the path is rooted (starts with a double colon `::`).
    pub rooted: bool,
    pub segments: heap::Vec<'heap, PathSegment<'heap>>,
}
