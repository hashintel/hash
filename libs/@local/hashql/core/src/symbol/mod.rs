//! Symbol representation and manipulation for HashQL.
//!
//! This module defines dedicated types for representing symbols in HashQL's compilation process.
//! Symbols are string-like values that appear in source code, such as identifiers, keywords,
//! and literals. They are deliberately kept separate from the syntax tree as they are used through
//! the different stages of compilation.
//!
//! The module provides:
//!
//! - [`Symbol`]: An opaque wrapper around string data that enables efficient storage and comparison
//! - [`Ident`]: A named identifier with source location and categorization
//! - [`IdentKind`]: Classification of different identifier types in HashQL
//!
//! ## Design Philosophy
//!
//! The [`Symbol`] type is designed as an opaque wrapper around its internal string storage.
//! This encapsulation enables future optimizations such as string interning (either through
//! the `string_interner` crate or a custom implementation) without requiring API changes.

pub mod sym;

use core::{
    borrow::Borrow,
    cmp::Ordering,
    fmt::{self, Display, Formatter},
    hash::{Hash, Hasher},
    ptr,
};

use crate::span::SpanId;

/// A string-like value used throughout the HashQL compiler.
///
/// Symbols represent string data that appears in source code and persists throughout
/// compilation, they are read-only and immutable.
///
/// This type is deliberately opaque to hide its internal representation,
/// allowing for future optimizations like string interning without changing
/// the public API. Symbols are designed to be efficient for long-lived objects
/// that are frequently compared, hashed, and referenced during compilation.
///
/// The caller must ensure that the string is unique and interned. The types correctness requires
/// relies on these *but it does not enforce it*.
#[derive(Debug, Copy, Clone)]
pub struct Symbol<'heap>(&'heap str);

impl<'heap> Symbol<'heap> {
    /// Creates a new interned symbol from a string slice.
    ///
    /// The caller must ensure that the string is unique and interned.
    pub(crate) const fn new_unchecked(string: &'heap str) -> Self {
        Self(string)
    }

    #[must_use]
    pub const fn as_str(&self) -> &str {
        self.0
    }

    /// Returns the string representation of the symbol.
    ///
    /// Unlike [`Self::as_str`], this method provides access for the lifetime of the interner
    /// instead of the symbol itself, somewhat circumventing the protections given to the symbol
    /// itself. Any unwrapped type should be considered no longer unique and interned.
    #[must_use]
    pub const fn unwrap(&self) -> &'heap str {
        self.0
    }

    #[must_use]
    pub const fn as_bytes(&self) -> &[u8] {
        self.0.as_bytes()
    }
}

impl AsRef<Self> for Symbol<'_> {
    #[inline]
    fn as_ref(&self) -> &Self {
        self
    }
}

impl PartialEq for Symbol<'_> {
    fn eq(&self, other: &Self) -> bool {
        // Pointer equality implies string equality (due to the unique contents assumption)
        ptr::eq(self.0, other.0)
    }
}

impl Eq for Symbol<'_> {}

impl PartialOrd for Symbol<'_> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Symbol<'_> {
    fn cmp(&self, other: &Self) -> Ordering {
        // Pointer equality implies string equality (due to the unique contents assumption), but if
        // not the same the contents must be compared.
        if self == other {
            Ordering::Equal
        } else {
            self.0.cmp(other.0)
        }
    }
}

impl Hash for Symbol<'_> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        // Pointer hashing is sufficient (due to the unique contents assumption)
        ptr::hash(self.0, state);
    }
}

impl Borrow<str> for Symbol<'_> {
    fn borrow(&self) -> &str {
        self.0
    }
}

impl AsRef<str> for Symbol<'_> {
    fn as_ref(&self) -> &str {
        self.0
    }
}

impl Display for Symbol<'_> {
    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
        Display::fmt(self.0, fmt)
    }
}

/// The classification of an identifier in HashQL.
///
/// HashQL supports different categories of identifiers, each with distinct
/// syntactic rules and semantic meanings in the language.
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum IdentKind {
    /// A lexical identifier following standard programming language naming rules.
    ///
    /// # Syntax
    ///
    /// Lexical identifiers in HashQL follow Rust's identifier parsing rules:
    /// - Begin with either an `XID_Start` Unicode character or an underscore (`_`)
    /// - Continue with zero or more `XID_Continue` Unicode characters
    /// - Cannot be a single underscore character (`_`)
    /// - Uppercase identifiers are used for types, while lowercase identifiers are used for values,
    ///   identifiers starting with an underscore (`_`) are considered lowercase.
    ///
    /// In formal notation:
    /// ```text
    /// IDENTIFIER_LEXICAL:
    ///     XID_Start XID_Continue*
    ///     | '_' XID_Continue+
    /// ```
    ///
    /// > Note: Zero-width characters are not allowed.
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// counter
    /// _identifier
    /// Москва        // Non-ASCII identifiers are supported
    /// 東京           // As are other scripts
    /// ```
    Lexical,

    /// A symbolic identifier consisting of operator-like characters.
    ///
    /// # Syntax
    ///
    /// Symbol identifiers can consist of ASCII operators or Unicode symbols/punctuation:
    ///
    /// ```text
    /// ASCII_SYMBOL:
    ///       '!' | '#' | '$' | '%' | '&'
    ///     | '*' | '+' | '.' | '/' | '<'
    ///     | '=' | '>' | '?' | '@' | '\'
    ///     | '^' | '|' | '-' | '~'
    ///
    /// RESERVED_SYMBOL:
    ///     ':'
    ///
    /// ; Excludes any symbols that are in the ASCII range
    /// UNICODE_SYMBOL:
    ///     Symbol
    ///     | Punctuation
    ///
    /// SYMBOL:
    ///     ASCII_SYMBOL
    ///     | UNICODE_SYMBOL
    ///
    /// IDENTIFIER_SYMBOL:
    ///     SYMBOL+
    ///     | '`' SYMBOL+ '`'
    /// ```
    ///
    /// > Note: `:` is reserved, and cannot be used as a symbol identifier.
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// <
    /// +
    /// ->
    /// ≥
    /// →
    /// `*`
    /// ```
    Symbol,

    /// An identifier representing a URL for external resource references.
    ///
    /// # Syntax
    ///
    /// Base URL identifiers must be valid `http` or `https` URLs that end with a forward slash
    /// (`/`), and must be enclosed within backticks:
    ///
    /// ```text
    /// IDENTIFIER_BASE_URL:
    ///     '`' HTTP_URL_STRING_ENDING_WITH_SLASH '`'
    /// ```
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// `https://blockprotocol.org/@blockprotocol/types/property-type/address/`
    /// `https://example.com/resources/schema/`
    /// `http://localhost:8080/api/v1/`
    /// ```
    BaseUrl,
}

/// A named identifier in HashQL with source location information.
///
/// An `Ident` combines a [`Symbol`] (the textual content of the identifier)
/// with a [`SpanId`] (its location in the source code) and a [`IdentKind`]
/// (its syntactic classification). Identifiers are used throughout the compiler
/// to represent variables, functions, types, and other named entities.
///
/// # Examples
///
/// In HashQL source code:
///
/// ```text
/// let counter = 0;
///     ^^^^^^^
/// ```
///
/// This identifier would be represented as an `Ident` with:
/// - name: `Symbol("counter")`
/// - kind: `IdentKind::Lexical`
/// - span: `5..12`
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Ident<'heap> {
    pub span: SpanId,

    pub value: Symbol<'heap>,
    pub kind: IdentKind,
}

#[cfg(test)]
impl<'heap> Ident<'heap> {
    pub(crate) const fn synthetic(value: Symbol<'heap>) -> Self {
        Self {
            span: SpanId::SYNTHETIC,
            value,
            kind: IdentKind::Lexical,
        }
    }
}

impl AsRef<str> for Ident<'_> {
    fn as_ref(&self) -> &str {
        self.value.as_str()
    }
}

impl Display for Ident<'_> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        Display::fmt(&self.value.0, fmt)
    }
}
