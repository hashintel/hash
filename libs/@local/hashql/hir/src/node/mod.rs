pub mod access;
pub mod branch;
pub mod call;
pub mod closure;
pub mod data;
pub mod graph;
pub mod input;
pub mod kind;
pub mod r#let;
pub mod operation;
pub mod variable;

use hashql_core::{id, span::SpanId};

use self::kind::NodeKind;

id::newtype!(
    /// A unique identifier for nodes in the HashQL High-Level Intermediate Representation (HIR).
    ///
    /// Each node in the HIR has a unique `NodeId` that can be used to track the node
    /// through various processing stages. These IDs are particularly important for
    /// error reporting, cross-referencing between nodes, and maintaining node identity
    /// during transformations.
    ///
    /// The value space is restricted to 0..=0xFFFF_FF00, reserving the last 256 for niches.
    /// As real pattern types are an experimental feature in Rust, these can currently only be
    /// used by directly modifying and accessing the `NodeId`'s internal value.
    pub struct HirId(u32 is 0..=0xFFFF_FF00)
);

impl HirId {
    /// A placeholder ID used during initial parsing.
    ///
    /// When parsing, we initially give all AST nodes this placeholder ID. After resolving special
    /// forms, the nodes are renumbered to have small indices.
    pub const PLACEHOLDER: Self = <Self as id::Id>::MAX;
}

/// A node in the HashQL High-Level Intermediate Representation (HIR).
///
/// The HIR is an optimized, more refined representation of the program derived from the AST.
/// While the AST closely mirrors the source syntax, the HIR represents a more semantically
/// meaningful structure that's better suited for type checking, optimization, and code generation.
///
/// Key differences between HIR and AST:
/// - Special forms (like `let`, `if`, etc.) are fully expanded in the HIR
/// - Syntactic sugar is desugared and normalized
/// - The HIR uses interned references for better memory efficiency
/// - Nodes are more compact and optimized for compiler operations
///
/// Each HIR node has:
/// - A unique identifier for tracking through compilation phases
/// - A span linking to its origin in source code for error reporting
/// - A specific kind that determines the node's semantics and structure
///
/// The HIR is designed to be immutable after construction and uses arena allocation
/// through the `'heap` lifetime for efficient memory management.
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct Node<'heap> {
    pub id: HirId,
    pub span: SpanId,

    pub kind: NodeKind<'heap>,
}
