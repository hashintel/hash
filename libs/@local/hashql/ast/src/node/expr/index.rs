use hashql_core::{heap, span::SpanId};

use super::Expr;
use crate::node::id::NodeId;

/// An indexing expression in the HashQL Abstract Syntax Tree.
///
/// Represents accessing an element from a collection by its index or key.
/// Indexing works with various collection types including lists, tuples,
/// dictionaries, structs, and other indexed data structures.
///
/// The index expression can be any expression that evaluates to a valid
/// index for the collection type - typically an integer for lists or a value of appropriate key
/// type for dictionaries.
///
/// Tuples and structs can be indexed using this expression, but only if the type system can prove
/// that the index is within bounds.
///
/// # Examples
///
/// ## J-Expr
///
/// ```json
/// ["[]", "items", 0]
/// ["items[0]"]
/// ["matrix[i][j]"]
/// ```
///
/// ## Documentation Format
///
/// ```text
/// items[0]
/// matrix[i][j]
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct IndexExpr<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub value: heap::Box<'heap, Expr<'heap>>,
    pub index: heap::Box<'heap, Expr<'heap>>,
}
