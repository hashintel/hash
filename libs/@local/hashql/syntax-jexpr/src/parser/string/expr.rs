use hashql_ast::node::{
    expr::{
        Expr, ExprKind, FieldExpr, IndexExpr, LiteralExpr,
        literal::{IntegerLiteral, LiteralKind},
    },
    id::NodeId,
};
use hashql_core::symbol::{Ident, IdentKind};
use winnow::{
    ModalResult, Parser as _,
    ascii::{digit1, multispace0},
    combinator::{
        alt, cut_err, delimited, dispatch, eof, fail, peek, preceded, repeat, terminated,
    },
    error::{AddContext, ParserError, StrContext, StrContextValue},
    token::any,
};

use super::{combinator::ws, context::Input, ident::parse_ident, path::parse_path};

#[derive(Debug)]
enum Access<'heap> {
    Index(LiteralExpr<'heap>),
    Field(Ident<'heap>),
}

// super limited set of expressions that are supported in strings for convenience
fn parse_field_access<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Access<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    preceded(
        ws("."),
        alt((
            parse_ident.map(Access::Field),
            digit1.with_span().map(|(digit, range)| {
                Access::Field(Ident {
                    span: context.span(range),

                    value: context.heap.intern_symbol(digit),
                    kind: IdentKind::Lexical, // Do we need to specify a different kind here?
                })
            }),
        )),
    )
    .parse_next(input)
}

fn parse_index_access<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Access<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    // super limited version that only allows literal access instead of arbitrary expressions
    delimited(
        ws("["),
        digit1.with_span().map(|(digit, range)| {
            let span = context.span(range);

            Access::Index(LiteralExpr {
                id: NodeId::PLACEHOLDER,
                span,
                kind: LiteralKind::Integer(IntegerLiteral {
                    value: context.heap.intern_symbol(digit),
                }),
                r#type: None,
            })
        }),
        ws(cut_err("]").context(StrContext::Expected(StrContextValue::CharLiteral(']')))),
    )
    .context(StrContext::Label("index"))
    .parse_next(input)
}

pub(crate) fn parse_expr_path<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Expr<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    (
        parse_path.with_span(),
        repeat(
            0..,
            preceded(
                multispace0,
                dispatch! {peek(any);
                    '[' => parse_index_access,
                    '.' => parse_field_access,
                    _ => fail
                        .context(StrContext::Expected(StrContextValue::CharLiteral('[')))
                        .context(StrContext::Expected(StrContextValue::CharLiteral('.')))
                }
                .with_span(),
            ),
        ),
    )
        .map(|((path, path_span), access): (_, Vec<_>)| {
            let mut range = path_span;

            let mut expr = Expr {
                id: NodeId::PLACEHOLDER,
                span: path.span,
                kind: ExprKind::Path(path),
            };

            for (access, span) in access {
                range.end = span.end;

                let span = context.span(range.clone());

                let kind = match access {
                    Access::Index(literal) => ExprKind::Index(IndexExpr {
                        id: NodeId::PLACEHOLDER,
                        span,
                        value: context.heap.boxed(expr),
                        index: context.heap.boxed(Expr {
                            id: NodeId::PLACEHOLDER,
                            span: literal.span,
                            kind: ExprKind::Literal(literal),
                        }),
                    }),
                    Access::Field(ident) => ExprKind::Field(FieldExpr {
                        id: NodeId::PLACEHOLDER,
                        span,
                        value: context.heap.boxed(expr),
                        field: ident,
                    }),
                };

                expr = Expr {
                    id: NodeId::PLACEHOLDER,
                    span,
                    kind,
                };
            }

            expr
        })
        .parse_next(input)
}

// super limited set of expressions that are supported in strings for convenience
pub(crate) fn parse_expr<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Expr<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    alt((
        terminated('_', eof).span().map(|span| Expr {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            kind: ExprKind::Underscore,
        }),
        parse_expr_path,
    ))
    .context(StrContext::Label("expression"))
    .parse_next(input)
}

#[cfg(test)]
mod tests {
    use super::{parse_expr, parse_field_access, parse_index_access};
    use crate::parser::string::test::{bind_parser, test_cases};

    // Bind our parsers to create testing functions
    bind_parser!(SyntaxDump; fn parse_expr_test(parse_expr));
    bind_parser!(Debug; fn parse_field_access_test(parse_field_access));
    bind_parser!(Debug; fn parse_index_access_test(parse_index_access));

    // Tests for field access
    test_cases!(parse_field_access_test;
        simple_field(".field") => "Simple field access",
        numeric_field(".123") => "Numeric field name",
        whitespace_before(".  field") => "Field access with whitespace before name",

        // Error cases
        missing_field_name(".") => "Missing field name",
        invalid_field_name(".@field") => "Invalid field name",
    );

    // Tests for index access
    test_cases!(parse_index_access_test;
        simple_index("[0]") => "Simple index access",
        multi_digit_index("[123]") => "Multi-digit index",
        whitespace_in_brackets("[ 42 ]") => "Whitespace inside brackets",

        // Error cases
        empty_brackets("[]") => "Empty brackets",
        unclosed_bracket("[42") => "Unclosed bracket",
        non_numeric_index("[abc]") => "Non-numeric index",
    );

    // Tests for full expressions
    test_cases!(parse_expr_test;
        // Simple paths
        simple_identifier("foo") => "Simple identifier",
        qualified_path("std::collections::HashMap") => "Qualified path",

        // Field access
        single_field_access("foo.bar") => "Single field access",
        chained_field_access("foo.bar.baz") => "Chained field access",
        numeric_field_access("foo.123") => "Numeric field access",

        // Index access
        single_index_access("foo[0]") => "Single index access",
        chained_index_access("foo[0][1]") => "Chained index access",

        // Mixed access
        mixed_access_1("foo.bar[0]") => "Field then index access",
        mixed_access_2("foo[0].bar") => "Index then field access",
        complex_access("foo.bar[0].baz[42].qux") => "Complex mixed access chain",

        // Path with generics
        generic_path("Vec<Int>") => "Path with generic",
        generic_with_access("Vec<Int>[0]") => "Generic path with index access",
        generic_with_field("Result<T, E>.value") => "Generic path with field access",

        // Whitespace handling
        whitespace_between_access("foo . bar [ 0 ]") => "Whitespace between access operations",
        whitespace_in_path("std :: collections :: HashMap") => "Whitespace in path",

        // Rooted paths
        rooted_path("::foo") => "Rooted path",
        rooted_with_access("::foo.bar[0]") => "Rooted path with access",

        // Underscore paths
        underscore("_") => "Underscore path",
        underscore_identifier("_name") => "Underscore identifier",
    );

    // Error cases for full expressions
    test_cases!(parse_expr_test;
        invalid_start(".foo") => "Starting with field access",
        invalid_start_2("[0]") => "Starting with index access",
        empty_expression("") => "Empty expression",
        unclosed_access_chain("foo.bar[0") => "Unclosed access chain",
        invalid_character("foo$bar") => "Invalid character in expression",
    );
}
