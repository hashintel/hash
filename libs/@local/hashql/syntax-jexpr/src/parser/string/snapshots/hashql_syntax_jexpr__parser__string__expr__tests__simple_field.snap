---
source: libs/@local/hashql/syntax-jexpr/src/parser/string/expr.rs
description: Simple field access
expression: ".field"
info:
  kind: Ok
---
Field(
    Ident {
        span: SpanId(
            1,
        ),
        name: Symbol(
            "field",
        ),
        kind: Lexical,
    },
)
